src/Pipeline.cc

```diff
@@ -58,6 +67,10 @@ Pipeline::popMe(const Http::StreamPointer &which)
     // in reality there may be multiple contexts doing processing in parallel.
     // XXX: pipeline still assumes HTTP/1 FIFO semantics are obeyed.
     assert(which == requests.front());
+    if (which->mayUseConnection()) {
+        Assure(requests.size() == 1);
+        poppedBusy = true;
```

Comment:
Does this code assume that after we pop a mayUseConnection() transaction, we will never push a regular GET transaction? I do not see poppedBusy becoming false after popMe(). Why is that assumption safe?! It feels like we could pop a POST transaction and then push a GET transaction.

Perhaps this code secretly relies on the being-popped completed POST transaction setting its mayUseConnection() to false before calling popMe()? In other words, perhaps this code relies on true mayUseConnection() during popMe() indicating some kind of "fatal" connection problem (i.e. a problem that is guaranteed to be handled by killing all transactions and closing the client-Squid connection so that we should not care about any future transactions on the same connection)?

Reply:
This PR should add explicit 'mayUseConnection(false)' in all required places. So calling popMe() on transaction which is still using connection means that there are(may) be still bytes belonging to that (bring removed) transaction. We cannot start another transaction with these bytes left behind (i.e., face message framing problems).

> this code relies on true mayUseConnection() during popMe() indicating some kind of "fatal" connection problem 

I think it is true, this poppedBusy flag indicates this. The second assertion in Pipeline::add() should uphold this invariant, i.e., we cannot add a transaction if there has been a 'fatal' error on the connection.